### PR TITLE
Remove writing through stream when caching data

### PR DESCRIPTION
**Pull request by markusthoemmes**
Having the streams in the deferred path hid a bug we had with XML handling, since they unescape when writing through them.

This moves the stream down to where it is actually needed: String based data handling. The data needs to be copied anyway, so this should also save a copy step.

Signed-off-by: Markus Thömmes <markusthoemmes@me.com>

## Benchmark

```
name                                                                                old time/op    new time/op    delta
Unmarshal/string_data_v0.3-16                                                         16.1µs ± 1%    15.7µs ± 1%  -2.01%  (p=0.008 n=5+5)
Unmarshal/nil_data_v0.3-16                                                            14.9µs ± 1%    14.7µs ± 0%  -1.48%  (p=0.008 n=5+5)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v0.3-16    16.9µs ± 0%    16.6µs ± 0%  -2.04%  (p=0.008 n=5+5)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v1.0-16    17.2µs ± 1%    16.7µs ± 0%  -2.95%  (p=0.008 n=5+5)
Unmarshal/string_data_v1.0-16                                                         15.9µs ± 1%    15.6µs ± 0%  -1.66%  (p=0.016 n=5+4)
Unmarshal/base64_json_encoded_data_v1.0-16                                            11.3µs ± 1%    11.2µs ± 1%  -1.14%  (p=0.008 n=5+5)
Unmarshal/base64_xml_encoded_data_v1.0-16                                             11.8µs ± 1%    11.6µs ± 1%  -1.71%  (p=0.008 n=5+5)
Unmarshal/struct_data_v0.3-16                                                         16.6µs ± 1%    16.2µs ± 1%  -2.00%  (p=0.008 n=5+5)
Unmarshal/nil_data_v1.0-16                                                            15.0µs ± 1%    14.9µs ± 1%  -1.05%  (p=0.032 n=5+5)
Unmarshal/struct_data_v1.0-16                                                         16.2µs ± 0%    16.0µs ± 1%    ~     (p=0.056 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
Unmarshal/string_data_v0.3-16                                                         3.36kB ± 0%    3.36kB ± 0%    ~     (all equal)
Unmarshal/nil_data_v0.3-16                                                            3.14kB ± 0%    3.14kB ± 0%    ~     (all equal)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v0.3-16    3.45kB ± 0%    3.40kB ± 0%  -1.48%  (p=0.008 n=5+5)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v1.0-16    3.45kB ± 0%    3.40kB ± 0%  -1.48%  (p=0.008 n=5+5)
Unmarshal/string_data_v1.0-16                                                         3.18kB ± 0%    3.18kB ± 0%    ~     (all equal)
Unmarshal/base64_json_encoded_data_v1.0-16                                            2.66kB ± 0%    2.66kB ± 0%    ~     (all equal)
Unmarshal/base64_xml_encoded_data_v1.0-16                                             2.67kB ± 0%    2.67kB ± 0%    ~     (all equal)
Unmarshal/struct_data_v0.3-16                                                         3.37kB ± 0%    3.37kB ± 0%    ~     (all equal)
Unmarshal/nil_data_v1.0-16                                                            3.14kB ± 0%    3.14kB ± 0%    ~     (all equal)
Unmarshal/struct_data_v1.0-16                                                         3.20kB ± 0%    3.20kB ± 0%    ~     (all equal)

name                                                                                old allocs/op  new allocs/op  delta
Unmarshal/string_data_v0.3-16                                                           67.0 ± 0%      67.0 ± 0%    ~     (all equal)
Unmarshal/nil_data_v0.3-16                                                              63.0 ± 0%      63.0 ± 0%    ~     (all equal)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v0.3-16      70.0 ± 0%      69.0 ± 0%  -1.43%  (p=0.008 n=5+5)
Unmarshal/data,_attributes_and_extensions_and_specversion_with_struct_data_v1.0-16      70.0 ± 0%      69.0 ± 0%  -1.43%  (p=0.008 n=5+5)
Unmarshal/string_data_v1.0-16                                                           65.0 ± 0%      65.0 ± 0%    ~     (all equal)
Unmarshal/base64_json_encoded_data_v1.0-16                                              48.0 ± 0%      48.0 ± 0%    ~     (all equal)
Unmarshal/base64_xml_encoded_data_v1.0-16                                               48.0 ± 0%      48.0 ± 0%    ~     (all equal)
Unmarshal/struct_data_v0.3-16                                                           68.0 ± 0%      68.0 ± 0%    ~     (all equal)
Unmarshal/nil_data_v1.0-16                                                              63.0 ± 0%      63.0 ± 0%    ~     (all equal)
Unmarshal/struct_data_v1.0-16                                                           66.0 ± 0%      66.0 ± 0%    ~     (all equal)
```

### CODE REVIEW

File: v2/event/event_unmarshal.go [hunk 1]
```diff
@@ -195,8 +176,18 @@ func ReadJson(out *Event, reader io.Reader) error {
...
+			// ... cache the data ...
+			if key == "data" || key == "data_base64" {
+				if key == "data_base64" {
+					appendFlag(&state, dataBase64Flag)
+				}
+				cachedData = iterator.SkipAndReturnBytes()
+				continue
+			}
+
+			// ... or enqueue for deferred handling.
```
[the remark this answers is not in the document]
Comment (markusthoemmes) [reply v2/event/event_unmarshal.go:190]: We're lacking a test then :(

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply v2/event/event_unmarshal.go:190]: Yep I think this particular corner case miss a test

[the remark this answers is not in the document]
Comment (markusthoemmes) [reply v2/event/event_unmarshal.go:190]: Fixed.

**Review by slinkydeveloper — APPROVED** [review]
